### ISSUE TITLE
AVIF compression options not working?

### ISSUE THREAD

**Issue by AbsurdePhoton** [Imagick/imagick#559]
Hi,

I posted this issue on ImagickDemos by mistake, I intended to post it here... sorry about that but I see no option to remove an issue?

I have this code to save an AVIF image :

```
if (class_exists('Imagick')) {
        $image = new Imagick();
        $image->readImage($file); // $file was defined before

        $image->setImageFormat('avif');
        //$image->setImageCompressionQuality($compression_quality); // defined before, default $compression_quality = 25
        //$image->setOption('heic:speed', 5);
        $image->writeImage($output);
    }
```

In the code I commented out the two options with which I tried to control the AVIF image quality. Whichever option I'm using, or both of them, the produced image size is the same, so **the compression options don't work**. I can successfully load the produced AVIF image under Chrome or Firefox.

I read the code here : https://github.com/ImageMagick/ImageMagick/blob/main/coders/heic.c - and it looks to me that both options should work?...

Versions used on my dev computer : Ubuntu 22.04, Imagemagick 7.1.0, Imagick 3.7.0, libheif 1.12

Thank you.

**Comment by Danack** [comment]
After spending an hour investigating...... `$image->setCompressionQuality` works, `$image->setImageCompressionQuality` doesn't.

Setting a quality of 100 appears to trigger lossless compression.

> but I see no option to remove an issue?

Yeah, there shouldn't be. People could abuse that by rapidly opening and removing issues.

btw do you have any idea what the 'speed' setting does?

**Comment by AbsurdePhoton** [comment]
Thanks for that tip, I'll try setCompressionQuality tonight.

As I understand it, the speed  option is the amount of computation used to refine the compression. But I didn't see any change when I used values from 0 to 10... And the overall compression speed should reflect this parameter... I suppose 10 is slower and 0 is faster?

Concerning the values, this is not clear. Quoted from https://github.com/libvips/libvips/issues/1788: 

> One thing to note however is that "speed" is not entirely straightforward as each libheif encoder plugin accepts different values; aom is 0-8 whereas rav1e is 0-10 whereas x265 is "superfast" to "veryslow".

See the code source here : https://github.com/strukturag/libheif/tree/master/libheif

I don't know which oneof the three decoders Imagick uses...

**Comment by Danack** [comment]
> I don't know which oneof the three decoders Imagick uses...

9 is the one that failed for me.

> aom is 0-8 whereas rav1e is 0-10 whereas x265 is "superfast" to "veryslow".

So "aom", then. Apparently.

Which does give me a feeling of being in a Monty Python sketch: https://www.dailymotion.com/video/x3v9z74

> But I didn't see any change when I used values from 0 to 10...

So.........one of the problems that HEIC/AVIF was apparently created to solve was banding in images.

TL:DR 8 bit colour images are enough for most things, except when you have large areas of smooth colour that transitions from one hue to another. And so HEIC/AVIF is meant to do special stuff (12 bit instead of 8, perceptual encoding) to avoid that problem

Code below produces these two images.....well. The code does produce two images but Github does not support avif images.

Opening them in Chrome on my computer it seems the speed 7 setting produces an image that looks better (to me) and takes less time to encode.

Which is interesting.


```
<?php

declare(strict_types=1);

function createGradientImage($width, $height, $colorPoints, $sparseMethod, $absolute = false)
{
    $imagick = new \Imagick();
    $imagick->newImage($width, $height, "white");
    $imagick->setImageFormat("png");

    $barycentricPoints = array();

    foreach ($colorPoints as $colorPoint) {
        if ($absolute == true) {
            $barycentricPoints[] = $colorPoint[0];
            $barycentricPoints[] = $colorPoint[1];
        } else {
            $barycentricPoints[] = $colorPoint[0] * $width;
            $barycentricPoints[] = $colorPoint[1] * $height;
        }

        if (is_string($colorPoint[2])) {
            $imagickPixel = new \ImagickPixel($colorPoint[2]);
        } else if ($colorPoint[2] instanceof \ImagickPixel) {
            $imagickPixel = $colorPoint[2];
        } else {
            $errorMessage = sprintf(
                "Value %s is neither a string nor an ImagickPixel class. Cannot use as a color.",
                $colorPoint[2]
            );

            throw new \InvalidArgumentException(
                $errorMessage
            );
        }

        $red = $imagickPixel->getColorValue(\Imagick::COLOR_RED);
        $green = $imagickPixel->getColorValue(\Imagick::COLOR_GREEN);
        $blue = $imagickPixel->getColorValue(\Imagick::COLOR_BLUE);
        $alpha = $imagickPixel->getColorValue(\Imagick::COLOR_ALPHA);

        $barycentricPoints[] = $red;
        $barycentricPoints[] = $green;
        $barycentricPoints[] = $blue;
        // TODO - document this as a change on IM7
        //$barycentricPoints[] = $alpha;
    }


    $imagick->sparseColorImage($sparseMethod, $barycentricPoints);

    return $imagick;
}


$points = [
    [0.30, 0.10, '#e700fc'],
    [0.10, 0.80, '#0400fc'],
    [0.70, 0.60, '#7b8896'],
    [0.80, 0.20, '#a9c9c4'],
];

$image = createGradientImage(
    800,
    800,
    $points,
    \Imagick::SPARSECOLORMETHOD_SPEPARDS
);


$image->setImageFormat('avif');
$compression = 100;

//for ($compression = 2; $compression < 100; $compression += 5) {
 for ($speed = 1; $speed < 8; $speed += 2) {
        //$speed = 5;
        echo "compression $compression speed $speed \n";
        $image->setCompressionQuality($compression); // defined before, default $compression_quality = 25

        $image->setOption('heic:speed', (string)$speed);
        $name = sprintf(
            "amagad_output_%d_%d.avif",
            $speed,
            $compression
        );
        $image->writeImage($name);
//  }
}
```

**Comment by Danack** [comment]
> I see you added a string cast when defining heic:speed. Maybe this is why it

No, I just default to using strict mode in PHP. If you pass an int it will be co-erced to an int internally.

> Speed set to 7 is faster, and the result is better?! Are you sure ? 

Yes. As I said....interesting.

The difference is tiny, but investigating the difference is not something I'm planning to do.

libheif is trying to encode the image so that it looks perceptually better, but that means that any of the steps of "image -> browser -> display adapter -> human eyes" could be the thing that is making a difference.

Which means I am completely unsure of how to check that it is behaving 'correctly'.

**Comment by AbsurdePhoton** [comment]
Thank you for investigating this, I was able to add AVIF images on my site, with real-time encoding of images.

Err, real-time not exactly anymore. I have a repository of photos in PNG format, and each time a new photo is displayed, a PHP function encodes it in several formats (JPEG, webP, AVIF) and several resolutions, on-the-fly. This done, I add an empty file to indicate the image was treated, so next time it isn't encoded.
With only JPEG and webP, it was reasonably fast.
BUT adding AVIF support added a _lot_  of waiting, several seconds for the large images on the main page which is a big carousel slide. With my photo galleries, some of them have 40+ photos, 1024 pixels on the largest side, so some pages hit the PHP time-out execution time. AVIF encoding is so SLOW. It's a known problem (https://avif.io/blog/articles/avif-faq/#howgoodisavifencodingspeed) and things should get better with time (optimizations, hardware support...) but for now it makes things harder to use.
So I ended up converting the biggest images on my site by hand, and uploading the result.

Thanks gain :)
